### PR TITLE
Ct sprint1

### PR DESCRIPTION
**Pull request by FelixChihTing**
This PR addresses [Issue#90](https://github.com/neurodata/brainlit/issues/90)

### CODE REVIEW

File: brainlit/algorithms/trace_analysis/fit_spline.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+"""
+Function definitions
+"""
+
+
+def checkIfDuplicates_2(listOfElems):
+    """Check if given list contains any duplicates
+    Args:
+        listOfElems (set): Build an unordered collection of unique elements.
```
Comment (tathey1) [inline brainlit/algorithms/trace_analysis/fit_spline.py]: I don't think the variable type is correct here. Also, the description seems unrelated.

Please read [this](https://google.github.io/styleguide/pyguide.html#s3.8.1-comments-in-doc-strings), and make sure docstrings adhere to this style.

File: brainlit/algorithms/trace_analysis/fit_spline.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+
+"""
+Geometric Graph class
+"""
+
+
+class GeometricGraph(nx.Graph):
+    r"""Class for undirected graphs.
+
```
Comment (tathey1) [inline brainlit/algorithms/trace_analysis/fit_spline.py:23]: Add information about why this class exists in the first place. It exists in order to fit splines to trees. It does this by, in part, requiring that each node has a loc attribute identifying that points location in space.

File: brainlit/algorithms/trace_analysis/fit_spline.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        # check there are no duplicate nodes
+        LOCs = [np.ndarray.tolist(self.nodes[node]["loc"]) for node in self.nodes]
+        LOCs.sort()
+        unique_LOCs = list(LOC for LOC, _ in itertools.groupby(LOCs))
+        if len(LOCs) != len(unique_LOCs):
+            raise ValueError("there are duplicate nodes")
+
+        # check the graph is edge-covering
+        if not nx.algorithms.is_edge_cover(self, self.edges):
```
Comment (tathey1) [inline brainlit/algorithms/trace_analysis/fit_spline.py:70]: edge_covering is not enough, you need the graph to be [connected](https://networkx.org/documentation/networkx-1.10/reference/generated/networkx.algorithms.components.connected.is_connected.html)

File: brainlit/algorithms/trace_analysis/fit_spline.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+            raise ValueError("the edges are not a valid cover of the graph")
+        # check there are no undirected cycles in the graph
+        if not nx.algorithms.tree.recognition.is_forest(self):
+            raise ValueError("the graph contains undirected cycles")
+        # check there are no disconnected segments
+        if not nx.algorithms.tree.recognition.is_tree(self):
+            raise ValueError("the graph contains disconnected segments")
```
Comment (tathey1) [inline brainlit/algorithms/trace_analysis/fit_spline.py:77]: why don't you do all of this in one test by testing if the graph is a [tree](https://networkx.org/documentation/stable//reference/algorithms/generated/networkx.algorithms.tree.recognition.is_tree.html)

File: brainlit/algorithms/trace_analysis/fit_spline.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        x = np.stack(x, axis=0)
+        new = x.shape[0]
+        if orig != new:
+            warnings.warn(
+                f"{orig-new} duplicate points removed in the trace segment",
+                category=UserWarning,
+            )
+        m = x.shape[0]
+        diffs = np.diff(x, axis=0)
+        diffs = np.linalg.norm(diffs, axis=1)
```
Comment (tathey1) [inline brainlit/algorithms/trace_analysis/fit_spline.py]: let's combine this into 1 line

File: brainlit/algorithms/trace_analysis/fit_spline.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+        diffs = np.cumsum(diffs)
+        diffs = np.concatenate(([0], diffs))
```
Comment (tathey1) [inline brainlit/algorithms/trace_analysis/fit_spline.py]: change to a more descriptive variable name

File: brainlit/algorithms/trace_analysis/fit_spline.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+        orig = x.shape[0]
+        x = [xi for i, xi in enumerate(x) if i == 0 or (xi != x[i - 1, :]).any()]
+        x = np.stack(x, axis=0)
+        new = x.shape[0]
+        if orig != new:
+            warnings.warn(
+                f"{orig-new} duplicate points removed in the trace segment",
+                category=UserWarning,
+            )
+        m = x.shape[0]
```
Comment (tathey1) [inline brainlit/algorithms/trace_analysis/fit_spline.py]: I think m is the path length right? If so, change variable name to "path_length"

File: brainlit/algorithms/trace_analysis/fit_spline.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        k = np.amin([m - 1, 5])
+        tck, u = splprep([x[:, 0], x[:, 1], x[:, 2]], u=diffs, k=k)
+
+        self.__check_multiplicity(tck[0])
+
+        return tck, u
+
+    def __check_multiplicity(self, t):
+        r"""Check multiplicity of elements in a list.
+
```
Comment (tathey1) [inline brainlit/algorithms/trace_analysis/fit_spline.py]: be more specific, it seems like you don't care about the multiplicity of the first and last knots, and that should be specified here.

File: docs/notebooks/utils/view_swc_spline.ipynb [hunk 9]
```diff
@@ -0,0 +1,203 @@
+{
```
Comment (tathey1) [inline docs/notebooks/utils/view_swc_spline.ipynb]: I will put comments for the notebook here:

- brainlit.algorithms.connect_fragments.fit_spline doesn't exist anymore
- Change "fit_spline_tree..." title to "Fitting Splines to Neuron Trace SWC Tutorial"
- Have the notebook read [this file](https://github.com/NeuroDataDesign/brainlit/tree/develop/data/data_octree/consensus-swcs) rather than "C:/Users/..."
- "extrated" typo
- After [this issue](https://github.com/neurodata/brainlit/issues/164) is addressed, please replace the code below 2) with a one-liner
- Make the neuron figure in 3) prettier (the red ball at the soma is obnoxious) and have a markdown cell explaining what's being shown 
- In 4), only plot a few branches, not every single one

**Review by tathey1 — CHANGES_REQUESTED** [review]

File: brainlit/algorithms/trace_analysis/fit_spline.py (hunk 1 above)
Comment (FelixChihTing) [reply brainlit/algorithms/trace_analysis/fit_spline.py]: I removed this function.

File: brainlit/algorithms/trace_analysis/fit_spline.py (hunk 2 above)
Comment (FelixChihTing) [reply brainlit/algorithms/trace_analysis/fit_spline.py:23]: I modified the [docstring](https://github.com/NeuroDataDesign/brainlit/blob/CTSprint1/brainlit/algorithms/trace_analysis/fit_spline.py#L21-L27).

File: brainlit/algorithms/trace_analysis/fit_spline.py (hunk 3 above)
Comment (FelixChihTing) [reply brainlit/algorithms/trace_analysis/fit_spline.py:70]: To check if the graph is edge-covering is not equivalent to check its connectivity. However, after passing the checks: ```is_edge_cover```, ```is_forest```, and ```is_tree```, we can determine why the graph is not a tree (containing cycles or disconnected). If the graph is disconnected, the graph will not pass the ```is_tree``` check.

File: brainlit/algorithms/trace_analysis/fit_spline.py (hunk 4 above)
Comment (FelixChihTing) [reply brainlit/algorithms/trace_analysis/fit_spline.py:77]: It's because we cannot directly test if an undirected graph contains cycles using the existed [algorithms](https://networkx.org/documentation/networkx-1.9/reference/algorithms.cycles.html). The advantage of this checking order is that we can determine why the graph is not a tree.

File: brainlit/algorithms/trace_analysis/fit_spline.py (hunk 7 above)
Comment (FelixChihTing) [reply brainlit/algorithms/trace_analysis/fit_spline.py]: Replaced ```m``` with ```path_length``` in [fit_spline.py](https://github.com/NeuroDataDesign/brainlit/blob/CTSprint1/brainlit/algorithms/trace_analysis/fit_spline.py#L148-L154).

File: brainlit/algorithms/trace_analysis/fit_spline.py (hunk 8 above)
Comment (FelixChihTing) [reply brainlit/algorithms/trace_analysis/fit_spline.py]: I removed ```__check_multiplicity```.

File: brainlit/algorithms/trace_analysis/fit_spline.py (hunk 5 above)
Comment (FelixChihTing) [reply brainlit/algorithms/trace_analysis/fit_spline.py]: Combined them into [1 line](https://github.com/NeuroDataDesign/brainlit/blob/CTSprint1/brainlit/algorithms/trace_analysis/fit_spline.py#L149).

File: docs/notebooks/utils/view_swc_spline.ipynb (hunk 9 above)
Comment (tathey1) [reply docs/notebooks/utils/view_swc_spline.ipynb]: dont worry about bullet 5

File: brainlit/algorithms/trace_analysis/fit_spline.py (hunk 6 above)
Comment (FelixChihTing) [reply brainlit/algorithms/trace_analysis/fit_spline.py]: already renamed:
the first ```diffs``` as ```NodeDist``` and the second ```diffs``` as ```TotalDist```.

File: docs/notebooks/utils/view_swc_spline.ipynb (hunk 9 above)
Comment (FelixChihTing) [reply docs/notebooks/utils/view_swc_spline.ipynb]: 1. path to ```fit_spline```modified
2. title changed
3. **not finished**
4. pyto fixed
5. skip
6. plots modified
7. reduce to 3 splines

Comment (FelixChihTing) [reply docs/notebooks/utils/view_swc_spline.ipynb]: 3. Added a [function](https://github.com/NeuroDataDesign/brainlit/blob/CTSprint1/docs/notebooks/algorithms/view_swc_spline.ipynb?short_path=927559d#L60-L81) to search for the swc file locally. It can fail when the current directory is lower than ```brainlit```folder.

**Review by tathey1 — APPROVED** [review]
nice work. I have a feeling that finding the swc file can be done more elegantly in the notebook, but this works.
